components/esp_http_server/src/httpd_uri.c

```diff
@@ -292,21 +293,33 @@ esp_err_t httpd_uri(struct httpd_data *hd)
     /* If URI with method not found, respond with error code */
     if (uri == NULL) {
         switch (err) {
-            case HTTPD_404_NOT_FOUND:
-                ESP_LOGW(TAG, LOG_FMT("URI '%s' not found"), req->uri);
-                return httpd_req_handle_err(req, HTTPD_404_NOT_FOUND);
-            case HTTPD_405_METHOD_NOT_ALLOWED:
-                ESP_LOGW(TAG, LOG_FMT("Method '%d' not allowed for URI '%s'"),
-                         req->method, req->uri);
-                return httpd_req_handle_err(req, HTTPD_405_METHOD_NOT_ALLOWED);
-            default:
-                return ESP_FAIL;
+        case HTTPD_404_NOT_FOUND:
+            ESP_LOGW(TAG, LOG_FMT("URI '%s' not found"), req->uri);
+            return httpd_req_handle_err(req, HTTPD_404_NOT_FOUND);
+        case HTTPD_405_METHOD_NOT_ALLOWED:
+            ESP_LOGW(TAG, LOG_FMT("Method '%d' not allowed for URI '%s'"),
+                     req->method, req->uri);
+            return httpd_req_handle_err(req, HTTPD_405_METHOD_NOT_ALLOWED);
+        default:
+            return ESP_FAIL;
         }
     }
 
     /* Attach user context data (passed during URI registration) into request */
     req->user_ctx = uri->user_ctx;
 
+    /* Final step for a WebSocket handshake verification */
+    if (uri->is_websocket && aux->ws_handshake_detect && uri->method == HTTP_GET) {
+        ESP_LOGD(TAG, LOG_FMT("Responding WS handshake to sock %d"), aux->sd->fd);
+        esp_err_t ret = httpd_ws_respond_server_handshake(&hd->hd_req);
+        if (ret != ESP_OK) {
+            return ret;
+        }
+
+        aux->sd->has_ws_handshake_done = true;
+        aux->sd->ws_handler = uri->handler;
+    }
+
     /* Invoke handler */
     if (uri->handler(req) != ESP_OK) {
```

Comment:
Is this supposed to be called even during the websocket handshake request? If yes, will it also get some data, or just the headers? If the handler is expected to parse headers, should it also get the facility of adding own headers ?

Reply:
@shahpiyushv All the WS requests that have been done handshaking before, will be handled separately before going in to the HTTP parser (or otherwise the parser will fail as it can't speak WebSockets anyway). Please refer to this function: https://github.com/huming2207/esp-idf/blob/feature/httpd-ws/components/esp_http_server/src/httpd_parse.c#L739